zend_abstract_interface/interceptor/php7/interceptor.c

```diff
@@ -0,0 +1,580 @@
+
+        // we do not use try / catch here as to preserve order of hooks, LIFO style, in bailout handler
+        if (prev) {
+            prev_execute_internal(execute_data, return_value);
+        } else {
+            func->internal_function.handler(execute_data, return_value);
+        }
+
+        zai_hook_finish(execute_data, return_value, &frame_memory.hook_data);
+        zai_hook_memory_table_del(execute_data);
+    } else {
+        if (prev) {
+            prev_execute_internal(execute_data, return_value);
+        } else {
+            func->internal_function.handler(execute_data, return_value);
+        }
+    }
+}
+
+static void zai_interceptor_execute_internal_no_prev(zend_execute_data *execute_data, zval *return_value) {
+    zai_interceptor_execute_internal_impl(execute_data, return_value, false);
+}
+
+static void zai_interceptor_execute_internal(zend_execute_data *execute_data, zval *return_value) {
+    zai_interceptor_execute_internal_impl(execute_data, return_value, true);
+}
+
+static zend_object_dtor_obj_t zai_interceptor_generator_dtor_obj;
+static void zai_interceptor_generator_dtor_wrapper(zend_object *object) {
+    zend_generator *generator = (zend_generator *)object;
+    zend_execute_data *execute_data = generator->execute_data;
+
+    zai_interceptor_generator_dtor_obj(object);
+
+    zai_interceptor_frame_memory *frame_memory;
+    if (execute_data && zai_hook_memory_table_find(execute_data, &frame_memory)) {
+        zval retval;
+        ZVAL_NULL(&retval);
+        if (!Z_ISUNDEF(generator->retval)) {
+            ZVAL_COPY_VALUE(&retval, &generator->retval);
+        }
+        zai_hook_finish(execute_data, &retval, &frame_memory->hook_data);
+        zai_hook_memory_table_del(execute_data);
+    }
+}
+
+static zend_object_handlers *zai_interceptor_generator_handlers;
+static pthread_once_t zai_interceptor_replace_generator_dtor_once = PTHREAD_ONCE_INIT;
+static void zai_interceptor_replace_generator_dtor(void) {
+    zai_interceptor_generator_dtor_obj = zai_interceptor_generator_handlers->dtor_obj;
+    zai_interceptor_generator_handlers->dtor_obj = zai_interceptor_generator_dtor_wrapper;
+}
+
+static zend_object *(*generator_create_prev)(zend_class_entry *class_type);
+static zend_object *zai_interceptor_generator_create(zend_class_entry *class_type) {
+    zend_generator *generator = (zend_generator *)generator_create_prev(class_type);
+#if PHP_VERSION_ID < 70100
+    zend_execute_data *execute_data = (zend_execute_data *)ZEND_VM_STACK_ELEMETS(EG(vm_stack));
+
+    zai_interceptor_frame_memory frame_memory;
+    if (zai_hook_continue(execute_data, &frame_memory.hook_data)) {
+        frame_memory.execute_data = execute_data;
+        zai_hook_memory_table_insert(execute_data, &frame_memory);
+    }
+#endif
+
+    zai_interceptor_generator_handlers = (zend_object_handlers *)generator->std.handlers;
+    pthread_once(&zai_interceptor_replace_generator_dtor_once, zai_interceptor_replace_generator_dtor);
```

Comment:
Compare this approach with:
https://github.com/DataDog/dd-trace-php/blob/e767a54f3113b2e886e07668f7749d079fa4549b/zend_abstract_interface/interceptor/php8/interceptor.c#L188-L205

I assume the other (in PHP 8 file) approach is probably a tiny bit safer and saving us the pthreads. We should decide for one specific approach across both versions though.
Other suggestions or comments welcome.